src/detection/scripts/ship_generator.py

```diff
@@ -0,0 +1,334 @@
+
+
+def check_in_bbox(x, y, bbox):
+    if (x > bbox[0] and x <= bbox[1]) and (y > bbox[2] and y <= bbox[3]):
+        return True
+    return False
+
+
+def contains(big, sml):
+    if big[0] <= sml[0] and big[1] <= sml[1] and big[2] >= sml[2] and \
+       big[3] >= sml[3]:
+        return True
+    return False
+
+
+def expand_window_with_offset(bbox):
+    """
+    This commented out portion is another method of generating offset,
+    with the new window location covering a larger area
+    box_w = abs(bbox[0] - bbox[1])
+    box_h = abs(bbox[2] - bbox[3])
+    center_x = (bbox[0] + bbox[1])/2
+    center_y = (bbox[2] + bbox[3])/2
+    valid_w = 512 - box_w
+    valid_h = 512 - box_h
+
+    ul_X = randrange(center_x - valid_w + 100, center_x + valid_w - 255)
+    ul_Y = randrange(center_y - valid_h + 100, center_y + valid_h - 255)
+    """
+
+    ul_X = randrange(bbox[0] - 100, bbox[0] + 100)
+    ul_Y = randrange(bbox[2] - 100, bbox[2] + 100)
+
+    return ul_X - 128, ul_X + 128, ul_Y - 128, ul_Y + 128
+
+
+def expand_window_no_offset(bbox):
+    if bbox[0] - 128 < 0 or bbox[0] + 128 < 0 or bbox[1] - 128 < 0 or \
+       bbox[1] + 128 < 0:
+        return 0, 256, 0, 256
+    return [bbox[0] - 128, bbox[0] + 128, bbox[2] - 128, bbox[2] + 128]
+
+
+def create_valid(chip_boxes, ship_boxes, maxX, maxY):
+    windows, ships = [], []
+    for chip in chip_boxes:
+        ships_in_chip = []
+        valid_window = True
+        if chip[0] < 0 or chip[1] < 0 or chip[2] > maxX or chip[3] > maxY:
+            valid_window = False
+        for ship in ship_boxes:
+            # Check if ship overlaps with the window and if it does
+            # check that the ship is completely in the window.
+            if not rasterio.coords.disjoint_bounds(chip, ship):
+                if not contains(chip, ship):
+                    valid_window = False
+                else:
+                    # Convert ship pixel location to location inside chip
+                    relative_ship = [ship[0] - chip[0], ship[2] - chip[0],
+                                     ship[1] - chip[1], ship[3] - chip[1]]
+                    ships_in_chip.append(relative_ship)
+        if valid_window:
+            windows.append(chip)
+            ships.append(ships_in_chip)
+
+    return windows, ships
+
+
+def generate_chips(data_type, viz):
+    from sklearn.preprocessing import MinMaxScaler
+    datasets_path = '/opt/Data/datasets/detection'
+    singapore_path = join(datasets_path, 'singapore')
+    chip_ships_list = []
+    chip_ind = 0
+
+    for img_ind, folder_name in enumerate(folders):
+        folder_path = join(singapore_path, folder_name) + '/'
+        img_name = data[img_ind][0]
+        img_corners = data[img_ind][1]
+        # src_path = folder_path + img_name                       # direct data
+        src_path = folder_path + 'test_' + str(img_ind) + '.tif'  # warped data
+        ogr_path = folder_path + 'ships_ogr_' + str(img_ind) + '.geojson'
```

Comment:
You should use `os.path.join` when building up paths instead of string concatenation.